Sources/AblyChat/ChatAPI.swift

```diff
@@ -0,0 +1,232 @@
+    private let rest: ARTRest
+    private let realtime: RealtimeClient
+    private let apiProtocolVersion: Int = 3
+    
+    init(rest: ARTRest, realtime: RealtimeClient) {
+        self.rest = rest
+        self.realtime = realtime
+    }
+
+    func getMessages(roomId: String, params: QueryOptions) async throws -> any PaginatedResult<Message> {
+        let endpoint = "/chat/v1/rooms/\(roomId)/messages"
+        let response: any PaginatedResult<Message> = try await makeAuthorizedPaginatedRequest(endpoint, params: params.toDictionary())
+        return response
+    }
+    
+    struct SendMessageResponse: Codable {
+        let timeserial: String
+        let createdAt: Int64
+    }
+
+    func sendMessage(roomId: String, params: SendMessageParams) async throws -> Message {
+        let endpoint = "/chat/v1/rooms/\(roomId)/messages"
+        var body: [String: Any] = ["text": params.text]
+        
+        if let metadata = params.metadata {
+            body["metadata"] = metadata
+        }
+        
+        if let headers = params.headers {
+            body["headers"] = headers
+        }
+        
+        let response: SendMessageResponse = try await makeAuthorizedRequest(endpoint, method: "POST", body: body)
+        let timeIntervalInSeconds = TimeInterval(integerLiteral: response.createdAt) / 1000
+        
+        let message = Message(
+            timeserial: response.timeserial,
+            clientID: realtime.clientId ?? "",
+            roomID: roomId,
+            text: params.text,
+            createdAt: Date(timeIntervalSince1970: timeIntervalInSeconds),
+            metadata: params.metadata ?? [:],
+            headers: params.headers ?? [:]
+        )
+        return message
+    }
+
+    func getOccupancy(roomId: String) async throws -> OccupancyEvent {
+        let endpoint = "/chat/v1/rooms/\(roomId)/occupancy"
+        return try await makeAuthorizedRequest(endpoint, method: "GET")
+    }
+    
+    private func makeAuthorizedRequest<RES: Codable>(_ url: String, method: String, body: [String: Any]? = nil) async throws -> RES {
+            return try await withCheckedThrowingContinuation { continuation in
+                do {
+                    try rest.request(method, path: url, params: [:], body: body, headers: ["protocol": String(apiProtocolVersion)], callback: { paginatedResponse, error  in
+                        
+                        if let error {
+                            continuation.resume(throwing: error)
+                            return
+                        }
+                        
+                        guard let firstItem = paginatedResponse?.items.first else {
+                            continuation.resume(throwing: ARTErrorInfo.createUnknownError())
+                            return
+                        }
+                        
+                        do {
+                            let jsonData = try JSONSerialization.data(withJSONObject: firstItem)
```

Comment:
This is an unfortunate dance (to JSON and back again). What can we do to avoid this? Create a decoder that directly decodes a dictionary? If not here, then I’d suggest creating an issue to return to.

Also, is it possible to reuse the `ARTHTTPPaginatedCallbackWrapper` type here to avoid having to write this logic in two places?

Reply:
is it better now?

Reply:
It's nice that the logic is in its own type, but the fact that we have to do a round-trip via JSON serialisation is still the same, and also is probably a bit confusing to someone looking at the code. As I said, I’d suggest creating an issue so we can revisit it.

Reply:
added a todo and opened an issue